### PR TITLE
Reorder the tasks executed by sozuctl upgrade

### PR DESCRIPTION
**Pull request by roosmaa**
Implements changes detailed in #282

### CODE REVIEW

**Comment by roosmaa** [comment]
There is one interesting observation: for a brief window new connections aren't accepted by sozu (connection refused). Is there a bug in sozu itself when it's executing it's upgrade or should the sozuctl run the upgrade commands in some other order?

**Comment by Geal** [comment]
Hello,
this will have to wait a bit, @NotBad4U is working on fixing soft stop and hard stop behaviours

**Comment by roosmaa** [comment]
Yup, understandable :)

**Comment by Geal** [comment]
handling new connections during the upgrade is something I'll address later, that feature is complicated to implement

File: ctl/src/main.rs [hunk 1]
```diff
@@ -27,11 +28,14 @@ fn main() {
   let config_file = matches.config;
 
   let config = Config::load_from_path(config_file.as_str()).expect("could not parse configuration file");
-  let mut channel: Channel<ConfigMessage,ConfigMessageAnswer> =
+  let connector = || -> Result<Channel<ConfigMessage,ConfigMessageAnswer>,io::Error> {
```
Comment (Geal) [inline ctl/src/main.rs]: I'm not a fan of this closure here. I'd rather have `upgrade` explicitely create the channel

File: ctl/src/command.rs [hunk 2]
```diff
@@ -505,7 +505,7 @@ pub fn status(mut channel: Channel<ConfigMessage,ConfigMessageAnswer>) {
   }
 }
 
-pub fn metrics(channel: &mut Channel<ConfigMessage,ConfigMessageAnswer>) {
+pub fn metrics(mut channel: Channel<ConfigMessage,ConfigMessageAnswer>) {
```
Comment (Geal) [inline ctl/src/command.rs]: please do not make unrelated changes when sending a pull request. Even if they can be small cosmetic changes, it makes the PR harder to figure out, and I have a limited time to review them.

File: ctl/src/command.rs [hunk 3]
```diff
@@ -247,92 +249,99 @@ pub fn upgrade<C>(mut channel: Channel<ConfigMessage,ConfigMessageAnswer>,
   ));
 
   match channel.read_message() {
-    None          => println!("the proxy didn't answer"),
+    None          => println!("Error: the proxy didn't list workers"),
```
Comment (Geal) [inline ctl/src/command.rs]: this case really means "the proxy did not answer", which is more information on the current state. Same for all other instances of this case afterwards

File: ctl/src/command.rs (hunk 2 above)
Comment (roosmaa) [reply ctl/src/command.rs]: Sure, I'll split that commit into a separate PR.

File: ctl/src/command.rs (hunk 3 above)
Comment (roosmaa) [reply ctl/src/command.rs]: I tried to find a balance between user friendliness and debuggability. Most of the time the those error messages tell little to the user, but having them be different should help in assisting finding the reason. If the proxy dies handling the n-th message it can be useful to know where to start the debugging from.

Though, if you think my reasoning above is not worth the changes, I can change it back to the "proxy didn't answer" message. Or to "proxy didn't answer [list workers message]"?

File: ctl/src/main.rs (hunk 1 above)
Comment (roosmaa) [reply ctl/src/main.rs]: Would a function `fn connect(config) -> Result<Channel<..>, io::Error>` be better? which means config instance would have to be passed into the upgrade command as such  `upgrade(connection, &config)`?

**Comment by Geal** [comment]
Thanks! I changed the code to use a function in main to create the channel, instead of a closure that's passed around
